### PR TITLE
Add Support for Foreign Assets

### PR DESCRIPTION
**Pull request by joepetrowski**
Adds support to Statemine and Westmint for foreign fungible assets. This will allow other parachains to represent their assets, as well as bridges to represent assets from other networks, on Statemine. A follow up PR will introduce this to Statemint once it has been audited and _in the Kusama coal mines_ for some time.

### CODE REVIEW

**Comment by bkontur** [comment]
bot rebase

**Comment by joepetrowski** [comment]
bot rebase

**Comment by joepetrowski** [comment]
bot rebase

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/statemint-update-roadmap/1200/94

**Comment by bkontur** [comment]
bot rebase

**Comment by bkontur** [comment]
bot rebase

**Comment by bkontur** [comment]
bot rebase

**Comment by bkontur** [comment]
bot rebase

**Comment by bkontur** [comment]
@joepetrowski 
TODOs from desc and all the rest is done in here https://github.com/paritytech/cumulus/pull/2167

**Comment by bkontur** [comment]
bot rebase

**Comment by joepetrowski** [comment]
bot rebase

File: parachains/runtimes/assets/common/src/lib.rs [hunk 1]
```diff
@@ -67,4 +107,201 @@ mod tests {
...
 		);
 	}
+
+	#[test]
+	fn trust_backed_assets_match_fungibles_works() {
+		frame_support::parameter_types! {
+			pub TrustBackedAssetsPalletLocation: MultiLocation = MultiLocation::new(0, X1(PalletInstance(13)));
+		}
+		// setup convert
+		type TrustBackAssetsConvert =
```
Comment (gavofyork) [inline parachains/runtimes/assets/common/src/lib.rs:117]:
```suggestion
		type TrustBackedAssetsConvert =
```
??

File: parachains/runtimes/assets/westmint/src/xcm_config.rs [hunk 2]
```diff
@@ -205,6 +243,34 @@ impl Contains<RuntimeCall> for SafeCallFilter {
 				pallet_assets::Call::touch { .. } |
 				pallet_assets::Call::refund { .. },
 			) |
+			RuntimeCall::ForeignAssets(
+				/* avoided: mint, burn */
```
Comment (gavofyork) [inline parachains/runtimes/assets/westmint/src/xcm_config.rs:247]: why avoid here? `mint` and `burn` permissions should never be available to any account anyway, right?

**Review by gavofyork — APPROVED** [review]
Looks like it should work from the runtime config. Didn't look closely at tests.

File: parachains/runtimes/assets/common/src/matching.rs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+impl<IsForeign: ContainsPair<MultiLocation, MultiLocation>> ContainsPair<MultiAsset, MultiLocation>
+	for IsForeignConcreteAsset<IsForeign>
+{
+	fn contains(asset: &MultiAsset, origin: &MultiLocation) -> bool {
+		log::trace!(target: "xcm::contains", "IsForeignConcreteAsset asset: {:?}, origin: {:?}", asset, origin);
+		matches!(asset.id, Concrete(ref id) if IsForeign::contains(id, &origin))
+	}
+}
+
+/// Checks if 'a' is from sibling location `b`, so means that `MultiLocation-a' starts with `MultiLocation-b'
```
Comment (bkchr) [inline parachains/runtimes/assets/common/src/matching.rs:68]: These docs are not correct :P 

I mean this is one part and then it also checks that it comes from a parachain that is a sibling.

**Review by bkchr — APPROVED** [review]
Looks good from a high level, haven't done an in depth review

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2562197

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2562727

File: parachains/runtimes/assets/westmint/src/xcm_config.rs [hunk 4]
```diff
@@ -100,8 +104,38 @@ pub type FungiblesTransactor = FungiblesAdapter<
...
+pub type ForeignFungiblesTransactor = FungiblesAdapter<
+	// Use this fungibles implementation:
+	ForeignAssets,
+	// Use this currency when it is a fungible asset matching the given location or name:
+	ForeignAssetsConvertedConcreteId,
+	// Convert an XCM MultiLocation into a local account id:
+	LocationToAccountId,
+	// Our chain's account ID type (we can't get away without mentioning it explicitly):
+	AccountId,
+	// TODO:check-parameter - no teleports
```
[the remark this answers is not in the document]
Comment (bkontur) [reply parachains/runtimes/assets/westmint/src/xcm_config.rs]: fixed with tests

**Comment by bkontur** [comment]
bot clean
bot rebase

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2578385

**Comment by bkontur** [comment]
bot merge

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-release-analysis-v0-9-41-v0-9-42/2828/1

**Pull request merged**